### ISSUE TITLE
Why not align with original implementation of authors in calculating so-called `pcon`

### ISSUE THREAD

**Issue by StepNeverStop** [RajGhugare19/dreamerv2#1]
https://github.com/RajGhugare19/dreamerv2/blob/65f5179b90720013fdbda6a332c2ba4a74ba7791/dreamerv2/training/trainer.py#L196

**Comment by RajGhugare19** [comment]
According to me "disc" in the original implementation and "discount_arr" in my implementation are not exactly the same.

Following image, from the original paper, describes the lambda-target used to train both actor and critic in imagination. Notice that according to the following image, at any time-index "t", disc[t] should contain the probability (ideally 0.99 if non terminating and 0 if terminating) of non-termination in the **state at "t+1"**.
![Screenshot from 2021-09-13 15-51-30](https://user-images.githubusercontent.com/62653460/133067625-e0ada6e4-4aeb-49eb-b147-a96ca7e1df25.png)

"discount_arr[t]" in my implementation contains the probability of non-termination at time-index "t", not "t+1" (see the commented [time-steps](https://github.com/RajGhugare19/dreamerv2/blob/ad7d54e256a03378bfd483e8a4c389ab2b444078/dreamerv2/training/trainer.py#L170) to understand how the tensors are stored). Hence, in my implementation after making the first row of discount_arr = ones, I concatenate it **with** discount_arr[1:] **and not with** discount_arr[:-1].  

```
#Pytorch Implementation (Mine):
discount_arr = torch.cat([torch.ones_like(discount_arr[:1]), discount_arr[1:]])
discount = torch.cumprod(discount_arr[:-1], 0)
```

According to this way of storing, my calculation of discount_arr (to be used  with actor and critic loss) seems to be logically coherent. Same thing is done in this [dreamer-v1 implementation](https://github.com/juliusfrost/dreamer-pytorch/blob/47bd509ab5cffa95ec613fd788d7ae1fe664ecd5/dreamer/algos/dreamer_algo.py#L276).

It could be possible that the original implementation is using slightly different storing mechanism, although I am not sure about it.

**Comment by HiddeLekanne** [comment]
First of, your results aren't correct, you're printing y for the tf.Tensor. You're results should be:

Tensorflow
```
tf.Tensor(
[[1.   1.   1.  ]
 [0.   0.1  0.2 ]
 [0.   0.04 0.1 ]], shape=(3, 3), dtype=float64)
```
Pytorch
```
tensor([[1.0000, 1.0000, 1.0000],
        [0.3000, 0.4000, 0.5000],
        [0.1800, 0.2800, 0.4000]], dtype=torch.float64)
```

Secondly it will bring different outcomes regardless  of it is sampled (1,0) or the mean <1,0>. 

And Thirdly, usually this difference has to do whether the states and the actions are (S_0, A_0) or (S_1, A_0) aligned, where S stands for State and A stands for A. So if this is indeed different for the different frameworks (like in one of the pytorch implementations of dreamer-v1) then they may actually not be different.
